axiom/nr_php_packages.c

```diff
@@ -0,0 +1,163 @@
+#include <stdlib.h>
+#include <stdio.h>
+
+#include "nr_axiom.h"
+#include "nr_limits.h"
+#include "nr_php_packages.h"
+#include "util_memory.h"
+#include "util_minmax_heap.h"
+#include "util_vector.h"
+#include "util_hashmap.h"
+#include "util_hashmap_private.h"
+#include "util_strings.h"
+
+nr_php_package_t* nr_php_package_create(char* name, char* version) {
+  nr_php_package_t* p = NULL;
+
+  if (NULL == name || NULL == version) {
+    return NULL;
+  }
+
+  p = (nr_php_package_t*)nr_malloc(sizeof(nr_php_package_t));
+  if (NULL == p) {
+    return NULL;
+  }
+
+  p->package_name = nr_strdup(name);
+  p->package_version = nr_strdup(version);
+
+  return p;
+}
+
+void nr_php_package_destroy(nr_php_package_t* p) {
+  if (NULL != p) {
+    nr_free(p->package_name);
+    nr_free(p->package_version);
+    nr_free(p);
+  }
+}
+
+void nr_php_packages_add_package(nr_php_packages_t** h, nr_php_package_t* p) {
+  nr_php_package_t* package;
+  if (NULL == h) {
+    return;
+  }
+
+  if (NULL == p || NULL == p->package_name || NULL == p->package_version) {
+    return;
+  }
+
+  if (NULL == *h) {
+    *h = nr_hashmap_create((nr_hashmap_dtor_func_t)nr_php_package_destroy);
+  }
+
+  // If package with the same key already exists, we will check if the value is
+  // different. If it is different, then we will update the value of the package
+  package = (nr_php_package_t*)nr_hashmap_get(*h, p->package_name,
+                                              nr_strlen(p->package_name));
+  if (NULL != package) {
+    if (0 != nr_strcmp(package->package_version, p->package_version)) {
+      nr_free(package->package_version);
+      package->package_version = nr_strdup(p->package_version);
+    }
+    nr_php_package_destroy(p);
+    return;
+  }
+
+  nr_hashmap_set(*h, p->package_name, nr_strlen(p->package_name), p);
+}
+
+void nr_php_packages_destroy(nr_php_packages_t** h) {
+  nr_hashmap_destroy(h);
+}
+
+size_t nr_php_packages_count(nr_php_packages_t* h) {
+  return nr_hashmap_count(h);
+}
+
+size_t nr_hashmap_count_buckets(const nr_php_packages_t* hashmap) {
+  return (size_t)(1 << hashmap->log2_num_buckets);
+}
+
+int nr_php_packages_has_package(nr_php_packages_t* h,
+                                char* package_name,
+                                size_t package_len) {
+  return nr_hashmap_has(h, package_name, package_len);
+}
+
+char* nr_php_package_to_json(nr_php_package_t* package) {
+  char* json;
+  nrbuf_t* buf = NULL;
+
+  if (NULL == package || NULL == package->package_name
+      || NULL == package->package_version) {
+    return NULL;
+  }
+  buf = nr_buffer_create(0, 0);
+  nr_buffer_add(buf, NR_PSTR("{\"name\":\""));
+  nr_buffer_add(buf, package->package_name, nr_strlen(package->package_name));
+  nr_buffer_add(buf, NR_PSTR("\",\"version\":\""));
+  nr_buffer_add(buf, package->package_version,
+                nr_strlen(package->package_version));
+  nr_buffer_add(buf, NR_PSTR("\"}"));
+  nr_buffer_add(buf, NR_PSTR("\0"));
```

Comment:
```suggestion
  nr_buffer_add(buf, NR_PSTR("[\""));
  nr_buffer_add(buf, package->package_name, nr_strlen(package->package_name));
  nr_buffer_add(buf, NR_PSTR("\",\""));
  nr_buffer_add(buf, package->package_version,
                nr_strlen(package->package_version));
  nr_buffer_add(buf, NR_PSTR("\""));

  /* unclear if this empty dictionary should be included or not */
  nr_buffer_add(buf, NR_PSTR(",{}]"));
  nr_buffer_add(buf, NR_PSTR("\0"));
```

Reply:
This change will generate JSON compatible with was is expected.

Reply:
You will need to fix unit tests which should be inspecting the generating JSON output.

Reply:
Thanks for fixing this! I have added these changes and updated the unit tests accordingly: [e997574](https://github.com/newrelic/newrelic-php-agent/pull/754/commits/e9975744e9d6ad61800b6952faafe32c74ec5dd9)